README.rst
========================
Ethiopian Date Converter
========================


.. image:: https://img.shields.io/pypi/v/ethiopian_date_converter.svg
        :target: https://pypi.python.org/pypi/ethiopian_date_converter

.. image:: https://img.shields.io/travis/jupi2142/ethiopian_date_converter.svg
        :target: https://travis-ci.org/jupi2142/ethiopian_date_converter

.. image:: https://readthedocs.org/projects/ethiopian-date-converter/badge/?version=latest
        :target: https://ethiopian-date-converter.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/jupi2142/ethiopian_date_converter/shield.svg
     :target: https://pyup.io/repos/github/jupi2142/ethiopian_date_converter/
     :alt: Updates


Ethiopian Date <==> Gregorian Date


* Free software: MIT license
* Documentation: https://ethiopian-date-converter.readthedocs.io.


Features
--------

* TODO

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage



HISTORY.rst
=======
History
=======

0.1.0 (2017-11-11)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for ethiopian_date_converter."""


tests/test_ethiopian_date_converter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `ethiopian_date_converter` package."""


import unittest

from ethiopian_date_converter import converter


class TestEthiopian_date_converter(unittest.TestCase):
    """Tests for `ethiopian_date_converter` package."""

    def test_ethiopic_to_gregorian(self):
        pairs = [
            [('5492-05-07', 'AA'), '0000-01-01'],
            [('5493-05-08', 'AA'), '0001-01-01'],
            [('5499-13-06', 'AA'), '0007-08-27'],
            [('5500-01-01', 'AA'), '0007-08-28'],
            [('5500-01-02', 'AA'), '0007-08-29'],
            [('5500-13-05', 'AA'), '0008-08-26'],
            [('1855-02-20', ), '1862-10-29'],
            [('1857-10-29', ), '1865-07-05'],
            [('1858-05-22', ), '1866-01-29'],
            [('1858-08-10', ), '1866-04-17'],
            [('1859-04-28', ), '1867-01-05'],
            [('1860-05-05', ), '1868-01-13'],
            [('0001-01-01', ), '0008-08-27'],
            [('0002-01-01', ), '0009-08-27'],
            [('0003-01-01', ), '0010-08-27'],
            [('0004-01-01', ), '0011-08-28'],
            [('0001-13-05', ), '0009-08-26'],
            [('0002-13-05', ), '0010-08-26'],
            [('0003-13-05', ), '0011-08-26'],
            [('0003-13-06', ), '0011-08-27'],
            [('0004-13-05', ), '0012-08-26'],
            [('1575-02-06', ), '1582-10-13'],
            [('1575-02-07', ), '1582-10-14'],
            [('1575-02-08', ), '1582-10-15'],
            [('1575-02-09', ), '1582-10-16'],
            [('1892-04-23', ), '1900-01-01'],
            [('1997-04-23', ), '2005-01-01'],
            [('2000-13-05', ), '2008-09-10'],
            [('1893-04-22', ), '1900-12-31'],
            [('1985-04-22', ), '1992-12-31'],
            [('1989-04-22', ), '1996-12-31'],
            [('1993-04-22', ), '2000-12-31'],
            [('1997-04-22', ), '2004-12-31'],
            [('2001-04-22', ), '2008-12-31'],
            [('2993-04-14', ), '3000-12-31'],
            [('3993-04-07', ), '4000-12-31'],
            [('5993-03-22', ), '6000-12-31']]

        for ethiopic, gregorian in pairs:
            assert converter(*ethiopic, to='gregorian') == gregorian, gregorian

    def test_gregorian_to_ethiopic(self):
        pairs = [[('1862-10-29', ), '1855-02-20'],
                 [('1865-07-05', ), '1857-10-29'],
                 [('1866-01-29', ), '1858-05-22'],
                 [('1866-04-17', ), '1858-08-10'],
                 [('1867-01-05', ), '1859-04-28'],
                 [('1868-01-13', ), '1860-05-05'],
                 [('0001-01-01', ), '5493-05-08'],
                 [('0007-08-27', ), '5499-13-06'],
                 [('0007-08-28', ), '5500-01-01'],
                 [('0007-08-29', ), '5500-01-02'],
                 [('0008-08-27', ), '0001-01-01'],
                 [('0009-08-27', ), '0002-01-01'],
                 [('0010-08-27', ), '0003-01-01'],
                 [('0011-08-28', ), '0004-01-01'],
                 [('0008-08-26', ), '5500-13-05'],
                 [('0009-08-26', ), '0001-13-05'],
                 [('0010-08-26', ), '0002-13-05'],
                 [('0011-08-26', ), '0003-13-05'],
                 [('0011-08-27', ), '0003-13-06'],
                 [('0012-08-26', ), '0004-13-05'],
                 [('1582-10-13', ), '1575-02-06'],
                 [('1582-10-14', ), '1575-02-07'],
                 [('1582-10-15', ), '1575-02-08'],
                 [('1582-10-16', ), '1575-02-09'],
                 [('1900-01-01', ), '1892-04-23'],
                 [('2005-01-01', ), '1997-04-23'],
                 [('2008-09-10', ), '2000-13-05'],
                 [('1900-12-31', ), '1893-04-22'],
                 [('1992-12-31', ), '1985-04-22'],
                 [('1996-12-31', ), '1989-04-22'],
                 [('2000-12-31', ), '1993-04-22'],
                 [('2004-12-31', ), '1997-04-22'],
                 [('2008-12-31', ), '2001-04-22'],
                 [('3000-12-31', ), '2993-04-14'],
                 [('4000-12-31', ), '3993-04-07'],
                 [('6000-12-31', ), '5993-03-22']]

        for gregorian, ethiopic in pairs:
            assert converter(*gregorian, to='ethiopian') == ethiopic, ethiopic

if __name__ == '__main__':
    unittest.main()


ethiopian_date_converter/ethiopian_date_converter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import calendar
from datetime import date
from functools import wraps

JD_EPOCH_OFFSET_AMETE_ALEM = -285019    # ዓ/ዓ
JD_EPOCH_OFFSET_AMETE_MIHRET = 1723856  # ዓ/ም
JD_EPOCH_OFFSET_GREGORIAN = 1721426

era_to_offset = {
    "AM": JD_EPOCH_OFFSET_AMETE_MIHRET,
    "AA": JD_EPOCH_OFFSET_AMETE_ALEM,
}


def string_dates(function):
    @wraps(function)
    def inner(date, *args, **kwargs):
        year, month, day = map(int, date.split('-'))
        return "{:04d}-{:02d}-{:02d}".format(
            *function(year, month, day, *args, **kwargs)
        )
    return inner


def ethiopian_coptic_to_julian_day_number(year, month, day, era):
    """
    Computes the Julian day number of the given Coptic or Ethiopic date.
    This method assumes that the JDN epoch offset has been set. This method
    is called by copticToGregorian and ethiopian_to_gregorian which will set
    the jdn offset context.

    :year: year in the Ethiopic calendar
    :month: month in the Ethiopic calendar
    :day: date in the Ethiopic calendar
    :era: [description]

    :returns: The Julian Day Number (JDN)
    """
    return (era + 365) + 365 * (year - 1) + (year / 4) + 30 * month + day - 31


def julian_day_number_to_gregorian(jdn,
                                   JD_OFFSET=JD_EPOCH_OFFSET_GREGORIAN,
                                   leapYear=calendar.isleap):
    """
    Converts JDN to Gregorian

    :jdn:
    :JD_OFFSET:
    :leapYear:
    :returns:
    """
    nMonths = 12
    monthDays = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

    r2000 = (jdn - JD_OFFSET) % 730485
    r400 = (jdn - JD_OFFSET) % 146097
    r100 = r400 % 36524
    r4 = r100 % 1461

    n = (r4 % 365) + 365 * (r4 / 1460)
    s = (r4 / 1095)
    aprime = 400 * ((jdn - JD_OFFSET) / 146097) + 100 * (r400 / 36524) + 4 * (
        r100 / 1461) + (r4 / 365) - (r4 / 1460) - (r2000 / 730484)
    year = aprime + 1
    t = ((364 + s - n) / 306)
    month = t * ((n / 31) + 1) + (1 - t) * (((5 * (n - s) + 13) / 153) + 1)
    n += 1 - (r2000 / 730484)
    day = n

    if ((r100 == 0) and (n == 0) and (r400 != 0)):
        month = 12
        day = 31
    else:
        monthDays[2] = 29 if (leapYear(year)) else 28
        for i in range(1, nMonths + 1):
            if n <= monthDays[i]:
                day = n
                break
            n -= monthDays[i]
    return (year, month, day)


def guessEra(jdn,
             JD_AM=JD_EPOCH_OFFSET_AMETE_MIHRET,
             JD_AA=JD_EPOCH_OFFSET_AMETE_ALEM):
    """
    Guesses ERA from JDN

    :jdn:
    :returns: {Number}
    """
    return JD_AM if jdn >= (JD_AM + 365) else JD_AA


def gregorian_to_julian_day_number(
        year=1, month=1,
        day=1, JD_OFFSET=JD_EPOCH_OFFSET_GREGORIAN):
    """
    Given year, month and day of Gregorian returns JDN

    :year:
    :month:
    :day:
    :JD_OFFSET:
    :returns:
    """
    s = (year / 4) - ((year - 1) / 4) - (year / 100) + (
        (year - 1) / 100) + (year / 400) - ((year - 1) / 400)
    t = ((14 - month) / 12)
    n = 31 * t * (month - 1) + (1 - t) * (59 + s + 30 * (month - 3) +
                                          ((3 * month - 7) / 5)) + day - 1
    j = JD_OFFSET + 365 * (year - 1) + ((year - 1) / 4) - ((year - 1) / 100) + (
        (year - 1) / 400) + n
    return j


def julian_day_number_to_ethiopic(jdn, era=JD_EPOCH_OFFSET_AMETE_MIHRET):
    """
    Given a JDN and an era returns the Ethiopic equivalent

    :jdn:
    :era:
    :returns: (year, month, day)
    """
    r = ((jdn - era) % 1461)
    n = (r % 365) + 365 * (r / 1460)
    year = 4 * ((jdn - era) / 1461) + (r / 365) - (r / 1460)
    month = (n / 30) + 1
    day = (n % 30) + 1
    return year, month, day


@string_dates
def ethiopian_to_gregorian(year=1,
                           month=1,
                           day=1,
                           era=JD_EPOCH_OFFSET_AMETE_MIHRET):
    if month not in range(1, 14):
        raise ValueError("Invalid month")
    if year <= 0:
        raise ValueError("Invalid year")
    if day not in range(1, 31):
        raise ValueError("Invalid day")
    if month == 13 and day not in range(1, 7):  # How about non leap year Pagume
        raise ValueError("Invalid day")
    return julian_day_number_to_gregorian(
        ethiopian_coptic_to_julian_day_number(year, month, day, era)
    )


@string_dates
def gregorian_to_ethiopic(year=1, month=1, day=1):
    date(year=year, month=month, day=day)
    jdn = gregorian_to_julian_day_number(year, month, day)
    return julian_day_number_to_ethiopic(jdn, guessEra(jdn))


def converter(date, era="AM", to='gregorian'):
    if to == 'gregorian':
        return ethiopian_to_gregorian(date, era=era_to_offset[era])
    elif to == 'ethiopian':
        return gregorian_to_ethiopic(date)
